### ISSUE TITLE
take advantage of stabilized cluster module

### ISSUE THREAD

**Issue by tjwebb** [trailsjs/trails#8]
Node `cluster` module is now set to Stable: https://nodejs.org/dist/latest-v4.x/docs/api/cluster.html

I think introducing clustering to the core would be useful and not too much work

**Comment by DaAwesomeP** [comment]
Trails also needs to be clusterable (is that a word?) by simple process replication and load balancing. Simply spawning more processes and equally directing traffic to it via a reverse proxy. This how most application hosters, Nginx (if you configure that way), PM2, and Passenger usually work.

**Comment by tjwebb** [comment]
>  This how most application hosters, Nginx (if you configure that way), PM2, and Passenger usually work

Yes, horizontal scalability definitely needs to be supported and configurable. That's not really related to this issue though. The `cluster` module can be used to utilize multiple processes on a single machine

**Comment by jacqueslareau** [comment]
Although adding basic clustering to a framework is trivial, I don't think it belongs there. I think process management should be a responsibility of a process manager. Something like PM2 has a lot a features to handle clustering, and trying to duplicate would be a waste.

With that being said, if the use case would be to enable clustering for a standalone app, it is very easy to find examples on how to use the cluster module. But I think the project should highly suggest using a proper process manager to handle processes.

**Comment by nrempel** [comment]
I agree with @jacqueslareau on this one.

**Comment by DaAwesomeP** [comment]
It depends on how the clustering is intended to work. If you're using to cluster module for simple horizontal scaling, than yes, it's useless. However, if you use it to aggregate tasks to different child processes specifically (such as background tasks), then it's great. It just depends on the level in which it's being used. If it is used to replicate the load-balancing function of a horizontal scale (i.e. with Nginx) where there is simply a base process that aggregates 100% of all and the exact same server task to different processes, then it is still useless, as Nginx is probably better at load balancing in this case.

Basically, it's good for the things that can't (or shouldn't) scale horizontally. This would include background jobs. I can't really think of much else that couldn't scale horizontally when combined with Redis.

**Comment by tjwebb** [comment]
The goal of this clustering work is not to replace an external process manager such as PM2. I intend and hope that trails will integrate nicely with tools like [forever](https://github.com/foreverjs/forever) and [pm2](https://github.com/Unitech/pm2).

Use of the [cluster](https://nodejs.org/api/cluster.html) module would be intended primarily to empower certain trailpacks to take advantage of the ability to split itself into multiple processes. For example, the hapi-trailpack may want to utilize multiple processes on a single machine.

I hope this clarifies the intention of this issue.

**Comment by DaAwesomeP** [comment]
@tjwebb I think that that `does` replace horizontal scaling, as the goal of horizontal scaling is to do exactly that. By allowing `hapi-trailpack` to do it's own load balancing to split to multiple processes, it is replicating the goal of horizontal scaling.

I understand your intention though. I just think that Hapi is a bad example. Background jobs in separate processes would definitely be a plus. However, there would need to be a semi-persistent queue (such as Redis) to ensure that the horizontally-scaled processes don't clash and spawn duplicates.

**Comment by JAertgeerts** [comment]
+1

**Comment by tjwebb** [comment]
Along the lines of @DaAwesomeP's point, I think the best way to enable this is to stay out of the way and let trailpacks implement this however they see fit.
